File: crates/poe-optimizer-pob/tests/support/skill_preparation_source.lua
-- Test-only UI boundary and observations around unchanged original methods.
-- This harness always runs complete Load, LoadSkill, ProcessSocketGroup,
-- FindSkillGem, SetDisplayGroup, and reached CalcTools helper bodies.
local function noop() end
local function dropdown(list)
 return setmetatable({list=list,selIndex=1},{__index=DropDownClass})
end
local function edit()
 return setmetatable({buf='',ResetUndo=noop,changeFunc=function()error('unexpected UI notification')end},{__index=EditClass})
end
local function scalar_fields(value)
 local result={}
 for key,item in pairs(value)do
  local kind=type(item)
  if kind=='boolean'or kind=='number'or kind=='string'then result[key]=item end
 end
 return result
end
local function load(xml,selection)
 local roots,parse_error=originalXml.ParseXML(xml);assert(roots,parse_error)
 local ordinal=0
 local function stamp(node)
  if type(node)~='table'then return end
  node._source_ordinal=ordinal;ordinal=ordinal+1
  for _,child in ipairs(node)do stamp(child)end
 end
 stamp(roots[1])
 local character_level=1
 for _,node in ipairs(roots[1])do
  if type(node)=='table'and node.elem=='Build'then character_level=tonumber(node.attrib.level)or 1 end
 end
 local build={data=data,characterLevel=character_level,SyncLoadouts=noop}
 local tab=setmetatable({build=build,controls={},ResetUndo=noop},{__index=SkillsTabClass})
 original_skill_defaults(tab)
 local lists=original_control_lists()
 tab.controls.defaultLevel=dropdown(lists.defaultGemLevelList)
 tab.controls.defaultQuality=edit()
 tab.controls.showSupportGemTypes=dropdown(lists.showSupportGemTypeList)
 tab.controls.sortGemsByDPSFieldControl=dropdown(lists.sortGemTypeList)
 tab.controls.sortGemsByDPS={};tab.controls.showLegacyGems={};tab.controls.groupList={}
 tab.controls.groupLabel=edit();tab.controls.groupSlot=dropdown(lists.groupSlotDropList)
 tab.controls.groupEnabled={};tab.controls.includeInFullDPS={};tab.controls.groupCount=edit()
 tab.UpdateGemSlots=function(self)
  self.gemSlots={}
  for index in ipairs(self.displayGroup.gemList)do
   self.gemSlots[index]={nameSpec=edit(),level=edit(),quality=edit(),enabled={},enableGlobal1={},enableGlobal2={},count=edit(),corruptLevel={}}
  end
 end
 local groups,seen,current={},{}
 tab.ProcessSocketGroup=function(self,group)
  local observed=seen[group]
  if not observed then
   assert(current,'unattributed group')
   observed={group=group,node=current,source=current._source_ordinal,attached=false,processing_passes=0}
   table.insert(groups,observed);seen[group]=observed
  end
  observed.processing_passes=observed.processing_passes+1
  -- Observation only: forward complete original processing and preserve errors.
  return SkillsTabClass.ProcessSocketGroup(self,group)
 end
 tab.LoadSkill=function(self,node,set_id)
  current=node
  local ok,err=pcall(SkillsTabClass.LoadSkill,self,node,set_id)
  for _,observed in ipairs(groups)do
   if observed.node==node and ok then observed.attached=true end
  end
  current=nil
  if not ok then error(err,0)end
 end
 local containers={}
 local function observe_load(node)
  local ok,err=pcall(SkillsTabClass.Load,tab,node,'independent skill preparation oracle')
  local fields={}
  for _,name in ipairs({'defaultGemLevel','defaultGemQuality','showSupportGemTypes','sortGemsByDPSField','sortGemsByDPS','showLegacyGems'})do fields[name]=tab[name]end
  table.insert(containers,{source=node._source_ordinal,fields=fields,order=copyTable(tab.skillSetOrderList),active_set_id=(ok or tab.activeSkillSetId~=0)and tab.activeSkillSetId or nil})
  if not ok then error(err,0)end
 end
 local ok,err=pcall(function()
  if roots[1].elem=='Skills'then
   observe_load(roots[1])
  else
   for _,node in ipairs(roots[1])do
    if type(node)=='table'and node.elem=='Skills'then observe_load(node)end
   end
  end
  if selection then tab:SetActiveSkillSet(selection)end
 end)
 local result={containers=containers,success=ok,error=not ok and tostring(err)or nil,groups={},selected_groups={},selected_key=tab.activeSkillSetId}
 for _,observed in ipairs(groups)do
  local record={source=observed.source,attached=observed.attached,processing_passes=observed.processing_passes,fields=scalar_fields(observed.group),gems={}}
  for index,gem in ipairs(observed.group.gemList)do
   local node=observed.node[index]
   table.insert(record.gems,{source=node._source_ordinal,fields=scalar_fields(gem),gem_data=gem.gemData and gem.gemData.id,granted_effect=gem.grantedEffect and gem.grantedEffect.id,stat_set=gem.statSet,stat_set_calcs=gem.statSetCalcs,minion_skill_lookup=gem.skillMinionSkillStatSetIndexLookup,minion_skill_lookup_calcs=gem.skillMinionSkillStatSetIndexLookupCalcs})
  end
  table.insert(result.groups,record)
 end
 if ok then
  for _,group in ipairs(tab.socketGroupList)do table.insert(result.selected_groups,assert(seen[group]).source)end
 end
 return result
end
return{load=load}

File: crates/poe-optimizer-pob/tests/support/skill_preparation_source.rs
//! Independent original-source skill preparation oracle. The executed source bodies
//! are verified against the pinned submodule; only UI controls and observations live here.
use mlua::{Function, HookTriggers, Lua, Table, Value, VmState};
use poe_optimizer_pob::source;
use std::{
    collections::BTreeMap,
    path::PathBuf,
    sync::OnceLock,
    time::{Duration, Instant},
};
pub fn repository() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../..")
}
fn sources() -> &'static BTreeMap<String, String> {
    static SOURCES: OnceLock<BTreeMap<String, String>> = OnceLock::new();
    SOURCES.get_or_init(|| {
        let mut paths = vec![
            "runtime/lua/xml.lua".into(),
            "src/Modules/Common.lua".into(),
            "src/Modules/Data.lua".into(),
            "src/Modules/CalcTools.lua".into(),
            "src/Classes/SkillsTab.lua".into(),
            "src/Classes/DropDownControl.lua".into(),
            "src/Classes/EditControl.lua".into(),
            "src/Data/Global.lua".into(),
            "src/Data/Gems.lua".into(),
            "src/Data/SkillStatMap.lua".into(),
            "src/Data/Assets.lua".into(),
            "src/Data/Skills/SkillAssets.lua".into(),
        ];
        for name in [
            "act_str", "act_dex", "act_int", "other", "minion", "spectre", "sup_str", "sup_dex",
            "sup_int",
        ] {
            paths.push(format!("src/Data/Skills/{name}.lua"));
        }
        paths
            .into_iter()
            .map(|path: String| {
                let text = source::read_verified_text(
                    &repository().join("vendor/path-of-building-poe2"),
                    &path,
                )
                .unwrap();
                (path, text)
            })
            .collect()
    })
}
fn original(path: &str) -> &'static str {
    &sources()[path]
}
fn section<'a>(text: &'a str, start: &str, end: &str) -> &'a str {
    assert_eq!(text.matches(start).count(), 1, "source start {start}");
    let from = text.find(start).unwrap();
    let to = from + text[from..].find(end).unwrap();
    &text[from..to]
}
pub struct Oracle {
    lua: Lua,
    helpers: Table,
}
impl Oracle {
    pub fn new(warm: bool) -> Self {
        let lua = Lua::new();
        lua.set_memory_limit(512 * 1024 * 1024).unwrap();
        let start = Instant::now();
        lua.set_hook(
            HookTriggers::new().every_nth_instruction(100_000),
            move |_, _| {
                if start.elapsed() > Duration::from_secs(90) {
                    Err(mlua::Error::RuntimeError(
                        "skill source oracle deadline exceeded".into(),
                    ))
                } else {
                    Ok(VmState::Continue)
                }
            },
        )
        .unwrap();
        lua.load(if warm { "jit.on()" } else { "jit.off()" })
            .exec()
            .unwrap();
        lua.load("t_insert=table.insert;t_remove=table.remove;m_min=math.min;m_max=math.max;m_floor=math.floor;SkillsTabClass={};DropDownClass={};EditClass={};main={defaultGemQuality=0};data={};calcLib={};observedSkillModules={}").exec().unwrap();
        for (path, start, end) in [
            (
                "src/Classes/SkillsTab.lua",
                "function SkillsTabClass:GetCorruptIndex(gemInstance)",
                "function SkillsTabClass:LoadSkill(",
            ),
            (
                "src/Classes/SkillsTab.lua",
                "function SkillsTabClass:SetDisplayGroup(socketGroup)",
                "function SkillsTabClass:AddSocketGroupTooltip(",
            ),
            (
                "src/Modules/Common.lua",
                "function sanitiseText(text)",
                "-- Convert int to 4 bytes string",
            ),
            (
                "src/Modules/Common.lua",
                "function copyTable(tbl, noRecurse)",
                "do\n\tlocal subTableMap",
            ),
            (
                "src/Modules/Common.lua",
                "function tableConcat(t1,t2)",
                "--- Simple table value equality",
            ),
            (
                "src/Modules/Common.lua",
                "function round(val, dec)",
                "--- Rounds down a number",
            ),
            (
                "src/Classes/SkillsTab.lua",
                "function SkillsTabClass:LoadSkill(node, skillSetId)",
                "function SkillsTabClass:Draw(",
            ),
            (
                "src/Classes/SkillsTab.lua",
                "function SkillsTabClass:FindSkillGem(nameSpec)",
                "-- Set the skill to be displayed/edited",
            ),
            (
                "src/Classes/SkillsTab.lua",
                "function SkillsTabClass:CreateSkillSet(skillSetId, title)",
                "-- Creates a new skill set with title",
            ),
            (
                "src/Classes/SkillsTab.lua",
                "function SkillsTabClass:SetActiveSkillSet(skillSetId, deferSync)",
                "-- Loop over all socket groups",
            ),
            (
                "src/Classes/DropDownControl.lua",
                "function DropDownClass:SelByValue(value, key)",
                "function DropDownClass:GetSelValue()",
            ),
            (
                "src/Classes/EditControl.lua",
                "function EditClass:SetText(text, notify)",
                "function EditClass:SetPlaceholder(",
            ),
            (
                "src/Modules/CalcTools.lua",
                "function calcLib.validateGemLevel(gemInstance)",
                "local typeExpressionStack",
            ),
            (
                "src/Modules/CalcTools.lua",
                "function calcLib.getGemStatRequirement(level, multi, isSupport)",
                "-- Build table of stats for the given skill instance statset",
            ),
        ] {
            lua.load(section(original(path), start, end))
                .set_name(format!("@{path}"))
                .exec()
                .unwrap();
        }
        lua.globals()
            .set(
                "LoadModule",
                lua.create_function(|lua, name: String| {
                    let path = format!("src/{name}.lua");
                    if name.starts_with("Data/Skills/") && name != "Data/Skills/SkillAssets" {
                        lua.globals()
                            .get::<Table>("observedSkillModules")?
                            .raw_push(path.clone())?;
                    }
                    let text = sources().get(&path).ok_or_else(|| {
                        mlua::Error::RuntimeError(format!("unapproved oracle dependency {path}"))
                    })?;
                    lua.load(text).set_name(format!("@{path}")).eval::<Value>()
                })
                .unwrap(),
            )
            .unwrap();
        let data_source = original("src/Modules/Data.lua");
        lua.load(original("src/Data/Global.lua"))
            .set_name("@src/Data/Global.lua")
            .exec()
            .unwrap();
        let assembly = format!(
            "{}\n{}\n{}",
            section(data_source, "local skillTypes = {", "local itemTypes = {"),
            section(
                data_source,
                "local function makeSkillMod(",
                "-----------------\n-- Common Data"
            ),
            section(data_source, "-- Load skills\n", "-- Load minions\n")
        );
        let assembly_helpers: Table = lua.load(format!("{assembly}\nreturn {{setupGem=setupGem,mod=makeSkillMod,flag=makeFlagMod,skill=makeSkillDataMod}}"))
            .set_name("@src/Modules/Data.lua")
            .eval()
            .unwrap();
        lua.globals()
            .set("originalAssemblyHelpers", assembly_helpers)
            .unwrap();
        let skills = original("src/Classes/SkillsTab.lua");
        lua.load(format!("function original_skill_defaults(self)\n{}\nend\nfunction original_control_lists()\n{}\nreturn {{groupSlotDropList=groupSlotDropList,defaultGemLevelList=defaultGemLevelList,showSupportGemTypeList=showSupportGemTypeList,sortGemTypeList=sortGemTypeList}} end",section(skills,"\tself.socketGroupList = { }","\t-- Set selector"),section(skills,"local groupSlotDropList = {","---@class SkillsTab:"))).exec().unwrap();
        let xml: Table = lua.load(original("runtime/lua/xml.lua")).eval().unwrap();
        lua.globals().set("originalXml", xml).unwrap();
        let helpers = lua
            .load(include_str!("skill_preparation_source.lua"))
            .eval()
            .unwrap();
        Self { lua, helpers }
    }
    pub fn observe(&self, xml: &str, selected: Option<f64>) -> Table {
        let call: Function = self.helpers.get("load").unwrap();
        call.call((xml, selected)).unwrap()
    }
    pub fn mutate(&self, mutation: &str) {
        self.lua.load(mutation).exec().unwrap();
    }
}
